File: 22FALL/331.verify-preorder-serialization-of-a-binary-tree.py
#
# @lc app=leetcode id=331 lang=python3
#
# [331] Verify Preorder Serialization of a Binary Tree
#

# @lc code=start
class Solution(object):
    def isValidSerialization(self, preorder):
        degree = 1
        i = 0
        while i<len(preorder):
            if preorder[i] == ',':
                i += 1
                continue
            degree -= 1
            if degree < 0:
                return False
            if preorder[i] == '#':
                i += 1
                continue
            while i < len(preorder) and preorder[i] not in ',#':
                i += 1
            degree += 2
        return degree == 0

# @lc code=end

File: VSCode_CPP/331.verify-preorder-serialization-of-a-binary-tree.cpp
/*
 * @lc app=leetcode id=331 lang=cpp
 *
 * [331] Verify Preorder Serialization of a Binary Tree
 */

// @lc code=start
class Solution {
public:
    bool isValidSerialization(string preorder) {
        int i = 0;
        int leaves = 1;
        while (i<preorder.size()) {
            if (leaves == 0) return false;
            if (preorder[i] == ',') i++;
            else if (preorder[i] == '#') {
                i++;
                leaves--;
            } else {
                while (i<preorder.size() && preorder[i] != ',') i++;
                leaves++;
            }
        }
        return leaves == 0;
    }
};
// @lc code=end

